test/test_cpp_api_parity.py

```diff
@@ -0,0 +1,282 @@
+      param->sizes().vec() == params2[param.key()].sizes().vec() &&
+      param->allclose(params2[param.key()]),
+      parity_test_error_msg, ": ", "sizes or value of `", param.key(), "` doesn't match");
+  }
+
+  auto buffers1 = m1->named_buffers();
+  auto buffers2 = m2->named_buffers();
+  TORCH_CHECK(
+    buffers1.size() == buffers2.size(),
+    parity_test_error_msg, ": ", "# of buffers doesn't match");
+  for (auto& buffer : buffers1) {
+    TORCH_CHECK(
+      buffer->sizes().vec() == buffers2[buffer.key()].sizes().vec() &&
+      buffer->allclose(buffers2[buffer.key()]),
+      parity_test_error_msg, ": ", "sizes or value of `", buffer.key(), "` doesn't match");
+  }
+}
+"""
+
+TORCH_NN_MODULE_TEST_INIT = Template("""\n
+void ${module_variant_name}_test_init(
+    const std::string& saved_module_path,
+    const std::string& device) {
+  ${module_qualified_name} m_init_by_python${cpp_constructor_args};
+  torch::load(m_init_by_python, saved_module_path);
+
+  torch::manual_seed(2);
+  ${module_qualified_name} m_init_by_cpp${cpp_constructor_args};
+  m_init_by_cpp->to(device);
+
+  test_module_state_equality(m_init_by_cpp.ptr(), m_init_by_python.ptr());
+}
+""")
+
+TORCH_NN_MODULE_TEST_FORWARD = Template("""\n
+void ${module_variant_name}_test_forward(
+    const std::string& saved_module_path,
+    const std::string& device,
+    torch::Tensor python_output,
+    ${input_arg_declarations}) {
+  torch::manual_seed(2);
+  ${module_qualified_name} module${cpp_constructor_args};
+  torch::load(module, saved_module_path);
+  module->to(device);
+
+  auto cpp_output = module(${input_args});
+
+  TORCH_CHECK(
+    cpp_output.sizes().vec() == python_output.sizes().vec() &&
+    cpp_output.allclose(python_output),
+    parity_test_error_msg, ": forward output doesn't match");
+}
+""")
+
+TORCH_NN_MODULE_TEST_BACKWARD = Template("""\n
+void ${module_variant_name}_test_backward(
+    const std::string& saved_module_path,
+    const std::string& saved_grad_module_path,
+    const std::string& device,
+    ${input_arg_declarations}) {
+  ${module_qualified_name} grad_module${cpp_constructor_args};
+  torch::load(grad_module, saved_grad_module_path);
+
+  torch::manual_seed(2);
+  ${module_qualified_name} module${cpp_constructor_args};
+  torch::load(module, saved_module_path);
+  module->to(device);
+
+  auto cpp_output = module(${input_args});
+  cpp_output.sum().backward();
+
+  for (size_t i = 0; i < module->parameters().size(); i++) {
+    auto named_param = module->named_parameters()[i];
+    auto grad = grad_module->parameters()[i];
+    TORCH_CHECK(
+      named_param->grad().allclose(grad),
+      parity_test_error_msg, ": ", "gradient value of `", named_param.key(), "` doesn't match");
+  }
+}
+""")
+
+class TestCppApiParity(common.TestCase):
+    def _test_torch_nn_module(self, test_params):
+        torch_nn_test_methods = ['init', 'forward', 'backward']
+
+        def test_init(device, python_module_class, python_constructor_args):
```

Comment:
nit: The naming of `test_init` is a little misleading because we're not actually testing it here, we're running the python init and returning some things. Maybe something like `get_args_for_init` or `setup_init_test`?

ditto for `test_forward` and `test_backward`.